llvm/lib/IR/Attributes.cpp

```diff
@@ -1800,14 +1801,20 @@ AttributeList::intersectWith(LLVMContext &C, AttributeList Other) const {
   if (*this == Other)
     return *this;
 
-  // At least for now, only intersect lists with same number of params.
-  if (getNumAttrSets() != Other.getNumAttrSets())
-    return std::nullopt;
-
   SmallVector<std::pair<unsigned, AttributeSet>> IntersectedAttrs;
-  for (unsigned Idx : indexes()) {
-    auto IntersectedAS =
-        getAttributes(Idx).intersectWith(C, Other.getAttributes(Idx));
+  SmallSet<unsigned, 8> AllIndexes{};
+  AllIndexes.insert(indexes().begin(), indexes().end());
+  AllIndexes.insert(Other.indexes().begin(), Other.indexes().end());
```

Comment:
Can we iterate over `AttributeList::index_iterator(std::max(getNumAttrSets(), Other.getNumAttrSets()))` instead?

Reply:
Yeah that seems cleaner.

Reply:
In retrospect not doing this may have been bugy, something like `foo(<must preserver> X)` could merge with `foo(X) <can drop>` Added a test for that case either way.